druid-shell/src/platform/x11/application.rs

```diff
@@ -350,20 +361,71 @@ impl Application {
         Ok(false)
     }
 
-    pub fn run(self, _handler: Option<Box<dyn AppHandler>>) {
+    fn run_inner(self) -> Result<(), Error> {
+        // Try to figure out the refresh rate of the current screen. We run the idle loop at that
+        // rate.
+        let refresh_rate = util::refresh_rate(self.connection(), self.window_id).unwrap_or(60.0);
+        let timeout = Duration::from_millis((1000.0 / refresh_rate) as u64);
+        let mut last_idle_time = Instant::now();
         loop {
-            if let Ok(ev) = self.connection.wait_for_event() {
+            let next_idle_time = last_idle_time + timeout;
+            poll_with_timeout(&self.connection, self.idle_read, next_idle_time)
+                .context("Error while waiting for X11 connection")?;
```

Comment:
Right now the order of the code here is the following (all wrapped in a big `loop { ... }`):

1. `poll` for readability
2. get events from x11rb until none are left in the queue
3. run idle handlers if there are some

If anything in step (3) causes x11rb to read an event from the X11 connection and enqueue it in its internal state, you could end up blocking in `poll`.

I propose to do something like this: https://github.com/psychon/x11rb/blob/87567eafa964efc6acd6bb7b76b69c1f1ff71215/examples/xeyes.rs#L308
Well... not really...

Anyway, what I mean:
Currently step (1) is `poll()` (pseudo-code). I propose to replace this with `let mut event = self.connection.poll_for_event(); if event.is_none() { poll(); } `.
Somehow you would then have to change step (2) so that this extra event is not lost.

Does this make some sense?

Reply:
That makes perfect sense, thanks!